File: apps/api/src/modules/products/product.controller.ts
import { Request, Response } from 'express';
import { productService } from './product.service.js';

export const productController = {
  async getAll(_req: Request, res: Response): Promise<void> {
    try {
      const products = await productService.getAll();
      res.json({ success: true, data: products });
    } catch (error) {
      res.status(500).json({
        success: false,
        error: { message: 'Failed to fetch products', code: 'INTERNAL_ERROR' },
      });
    }
  },

  async getById(req: Request, res: Response): Promise<void> {
    try {
      const id = String(req.params['id']);
      const product = await productService.getById(id);
      
      if (!product) {
        res.status(404).json({
          success: false,
          error: { message: 'Product not found', code: 'NOT_FOUND' },
        });
        return;
      }
      
      res.json({ success: true, data: product });
    } catch (error) {
      res.status(500).json({
        success: false,
        error: { message: 'Failed to fetch product', code: 'INTERNAL_ERROR' },
      });
    }
  },

  async create(req: Request, res: Response): Promise<void> {
    try {
      const product = await productService.create(req.body);
      res.status(201).json({ success: true, data: product });
    } catch (error) {
      res.status(500).json({
        success: false,
        error: { message: 'Failed to create product', code: 'INTERNAL_ERROR' },
      });
    }
  },

  async update(req: Request, res: Response): Promise<void> {
    try {
      const id = String(req.params['id']);
      const product = await productService.update(id, req.body);
      
      if (!product) {
        res.status(404).json({
          success: false,
          error: { message: 'Product not found', code: 'NOT_FOUND' },
        });
        return;
      }
      
      res.json({ success: true, data: product });
    } catch (error) {
      res.status(500).json({
        success: false,
        error: { message: 'Failed to update product', code: 'INTERNAL_ERROR' },
      });
    }
  },

  async delete(req: Request, res: Response): Promise<void> {
    try {
      const id = String(req.params['id']);
      const deleted = await productService.delete(id);
      
      if (!deleted) {
        res.status(404).json({
          success: false,
          error: { message: 'Product not found', code: 'NOT_FOUND' },
        });
        return;
      }
      
      res.status(204).send();
    } catch (error) {
      res.status(500).json({
        success: false,
        error: { message: 'Failed to delete product', code: 'INTERNAL_ERROR' },
      });
    }
  },
};

File: micros/product-service/src/main/java/com/qeetmart/product/controller/ProductController.java
package com.qeetmart.product.controller;

import com.qeetmart.product.dto.request.ProductCreateRequest;
import com.qeetmart.product.dto.request.ProductUpdateRequest;
import com.qeetmart.product.dto.response.ApiResponse;
import com.qeetmart.product.dto.response.PagedResponse;
import com.qeetmart.product.dto.response.ProductResponse;
import com.qeetmart.product.entity.ProductStatus;
import com.qeetmart.product.service.ProductService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/products")
@RequiredArgsConstructor
@Tag(name = "Product APIs", description = "Product management endpoints")
public class ProductController {

    private final ProductService productService;

    @PostMapping
    @PreAuthorize("hasRole('ADMIN')")
    @Operation(summary = "Create product")
    public ResponseEntity<ProductResponse> createProduct(@Valid @RequestBody ProductCreateRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(productService.createProduct(request));
    }

    @GetMapping
    @Operation(summary = "Get products with pagination and filters")
    public ResponseEntity<PagedResponse<ProductResponse>> getProducts(
        @RequestParam(required = false) Long categoryId,
        @RequestParam(required = false) String brand,
        @RequestParam(required = false) ProductStatus status,
        @PageableDefault(size = 10) Pageable pageable
    ) {
        return ResponseEntity.ok(productService.getProducts(categoryId, brand, status, pageable));
    }

    @GetMapping("/{productId}")
    @Operation(summary = "Get product by ID")
    public ResponseEntity<ProductResponse> getProductById(@PathVariable Long productId) {
        return ResponseEntity.ok(productService.getProductById(productId));
    }

    @PutMapping("/{productId}")
    @PreAuthorize("hasRole('ADMIN')")
    @Operation(summary = "Update product")
    public ResponseEntity<ProductResponse> updateProduct(
        @PathVariable Long productId,
        @Valid @RequestBody ProductUpdateRequest request
    ) {
        return ResponseEntity.ok(productService.updateProduct(productId, request));
    }

    @DeleteMapping("/{productId}")
    @PreAuthorize("hasRole('ADMIN')")
    @Operation(summary = "Soft delete product")
    public ResponseEntity<ApiResponse> deleteProduct(@PathVariable Long productId) {
        productService.deleteProduct(productId);
        return ResponseEntity.ok(ApiResponse.builder().message("Product deleted successfully").build());
    }

    @GetMapping("/search")
    @Operation(summary = "Search products by name or brand")
    public ResponseEntity<List<ProductResponse>> searchProducts(@RequestParam String query) {
        return ResponseEntity.ok(productService.searchProducts(query));
    }
}
